client/signup/config/flows.js

```diff
@@ -52,7 +52,18 @@ function getPostsDestination( dependencies ) {
 	return '/posts/' + dependencies.siteSlug;
 }
 
-const flows = generateFlows( { getPostsDestination, getSiteDestination } );
+function getRedirectDestination( dependencies ) {
+	if (
+		dependencies.redirect &&
+		dependencies.redirect.match( /^https%3A%2F%2Fpublic-api.wordpress.com/i )
```

Comment:
Shouldn't this check `dependencies.oauth2_redirect`? Also, the fact that the URL is encoded is suspicious. We use it later by assigning directly to `window.location.href`. And that doesn't accept an encoded URL. When I do this in console:
```js
window.location.href = "https%3A%2F%2Fwordpress.com";
```
the browser interprets the string as a relative path rather than full absolute URL.

Reply:
I'll follow up with a better approach and a server side patch. Lets merge this now to fix the bug...

Reply:
@jsnajdr that's a good spot and is definitely worth investigating! Just FYI I approved that PR since the fix actually works and doesn't modify the value that's getting passed to `window.location.href` which means we should be able to address the encoded vs unencoded value in a separate code diff

Reply:
Interesting 🙂 Let's figure out a better solution.

The login page redirects only to URLs that were first sanitized by a REST API call (i.e., returned as a response field) rather than using the one from query params directly.

The signup page doesn't seem to do that (yet).

Reply:
https://github.com/Automattic/wp-calypso/pull/32387